### ISSUE TITLE
cargo

### ISSUE THREAD

**Issue by geraldstanje** [japaric/ruststrap#35]
hi,

how did you install cargo?

**Comment by japaric** [comment]
If you want to use the existing (old) [cargo nightlies](https://www.dropbox.com/sh/qfbt03ys2qkhsxs/AACxFoD1OrxDXURzj5wX0IYUa?dl=0), I recommend using [multirust](https://github.com/japaric/ruststrap/blob/master/0-README.md#installation)

If you want a newer cargo, I recommend using an old ARM nightly to build a new one (in the ARM device). Use the normal `configure && make` procedure but pass the `--local-cargo=/path/to/cargo` to the `configure` script.

**Comment by geraldstanje** [comment]
or build it from source using:
curl -sSf https://static.rust-lang.org/rustup.sh | sh -s -- --prefix tmp ?

**Comment by geraldstanje** [comment]
do you talk about the configure script from here https://github.com/rust-lang/rust ?

**Comment by japaric** [comment]
You are trying to build a newer cargo on your ARM device, right? Because you mention that you already cross compiled a rust compiler to ARM, and you can install one of my older cargo nightlies with multirust. Then, something like this should work:

```
# On your ARM device
# `rustc` should already work
$ rustc -V
rustc 1.2.0-nightly (0cc99f9cc 2015-05-17) (built 2015-05-18)

# The old `cargo` too
$ cargo -V
cargo 0.2.0-nightly (e3a6425 2015-04-27) (built 2015-04-28)

# Get cargo source
$ git clone https://github.com/rust-lang/cargo

# configure
$ cd cargo
$ ./configure --enable-nightly --enable-optimize --local-cargo=/some/path/bin/cargo
(...)

# then make
$ make
(...)

# The output will be at target/arm-*/release
$ ls target/arm-unknown-linux-gnueabihf/release
(..)  cargo  (..)  libcargo-afd1dac7f2432d5a.rlib  (..)
```

**Comment by Arnold1** [comment]
nope, i built cargo on my ubuntu bc i cross compile for arm...

**Comment by japaric** [comment]
@Arnold1 The rustup.sh script installs a pre-compiled package that includes rust and cargo, and there are only precompiled packages for *officially supported* targets like linux, mac and windows on x86_64/i686. That script *won't* work on your ARM device, because the arm-linux-gnueabihf target is not officially supported.

**Comment by Arnold1** [comment]
i dont run cargo on the arm! cargo is used on ubuntu (x86_64)

**Comment by japaric** [comment]
@Arnold1 Can you clarify what you intend to do? I see that you have both rust and cargo on a x86_64 machine, what is the role of the ARM device? Do you want to cross compile a crate on the x86_64 machine, and run the produced binary on the ARM device?

**Comment by Arnold1** [comment]
yes exactly.

**Comment by japaric** [comment]
> i use ubuntu (x86_64) to cross compile rust for ARM, scp the binary to my ARM device and execute it there.

Yes, that the main idea. You may also need some libraries in the ARM device as well. Calling `ldd` on the ARM binary in your x86_64 machine may tell you which libraries you'll need in the ARM device. I've written about cross compiling in this [document](https://github.com/japaric/ruststrap/blob/master/1-how-to-cross-compile.md).

@Arnold1 or @geraldstanje  Could you tell me what specific problems are you encountering? A lot of things can go wrong when cross compiling, specially when C dependencies/libraries are involved.

**Comment by Arnold1** [comment]
hi, i have some problem when i build my app with:
```
cargo build --target=arm-unknown-linux-gnueabihf --release
```

i get an runtime error: illegal instruction

**Comment by japaric** [comment]
@Arnold1 

> i get a runtime error: illegal instruction

That may be caused because you compiled the rust program for armv7, or one of the C dependencies was compiled for armv7 but you are running the binary on an armv6 device.

Could you give me details about the device? Is an ARMv6 processor? Also what `rustc`/`cargo` version are you using? Did you download one of my unofficial tarballs, or did you compile it yourself from source?

**Comment by Arnold1** [comment]
here some infos:

rust:
```
rustc --version
rustc 1.0.0-dev (a59de37e9 2015-05-13) (built 2015-05-21)
```

cargo:
```
./cargo --version
cargo 0.2.0-nightly (efb482d 2015-04-30) (built 2015-04-30)
```
arm device:
```
armv7l
```

**Comment by japaric** [comment]
> i built rustc (cross compiler for arm) using your instruction

Here may be the problem. which gcc cross compiler did you use to cross compile the standard crates? The `g++-arm-linux-gnueabihf` package provided by Ubuntu? If so, which Ubuntu release are you using? And what's the output of `arm-linux-gnueabihf-gcc -v`?

Also what linux distribution are you running on the ARM device? Does your cargo crate have any C dependencies? Does a simple "Hello, world" program work fine on the ARM device?

By the way, I've uploaded [new tarballs for 1.0.0 and the nightly channel](https://github.com/japaric/ruststrap#nightlies), you may want to try those as well.

**Comment by geraldstanje** [comment]
i used:
```
[target.arm-unknown-linux-gnueabihf]
linker = "arm-linux-gnueabihf-gcc"
```

my output of arm-linux-gnueabihf-gcc -v:
```
gcc version 4.8.2 (Ubuntu/Linaro 4.8.2-16ubuntu4)
```

i used valgrind to see if there are any leaks:
- valgrind --log-file=out1.txt --leak-check=yes ./rust_app 100
http://paste.ubuntu.com/11482218
- valgrind --log-file=out1.txt --track-origins=yes ./rust_app 100
http://paste.ubuntu.com/11482231

**Comment by Arnold1** [comment]
i used gcc:
```
[target.arm-unknown-linux-gnueabihf]
linker = "arm-linux-gnueabihf-gcc"
```

my output of arm-linux-gnueabihf-gcc -v:
```
gcc version 4.8.2 (Ubuntu/Linaro 4.8.2-16ubuntu4)
```

i used valgrind to see if there are any leaks:
- valgrind --log-file=out1.txt --leak-check=yes ./rust_app 100
http://paste.ubuntu.com/11482218
- valgrind --log-file=out1.txt --track-origins=yes ./rust_app 100
http://paste.ubuntu.com/11482231

my cargo dependencies:
```
[dependencies]
libc = "*"
mmap = "*"
```
yes, simple hello world works fine...

on the arm device i run:
```
root_tar=ubuntu-core-14.04.2-core-armhf.tar.gz
root_url=http://cdimage.ubuntu.com/ubuntu-core/releases/14.04/release/$root_tar
```

**Comment by japaric** [comment]
If "hello world" works and your crate doesn't, could you try to generate a minimal working example that reproduces the "illegal instruction" problem, and share that with me?

**Comment by Arnold1** [comment]
crate also does work... but i implemented a function which doesnt work with --release. it works fine with debug.

yes here the function:
http://ideone.com/OjuSb9

when i change from:
```
(*self.chaMem)[i] = (*data)[i] as u32;
```

to:
```
(*self.chaMem)[i] = 12;
``

the code also works with --release
very strange...

**Comment by Arnold1** [comment]
i took the original code of the function and put a #[inline(never)] before ... code runs fine :D

**Comment by japaric** [comment]
If it works in debug but not under release, that could indicate that you are relying on undefined behavior at some point.

> from ... to ...

That sounds like `data` may have been freed/unmapped at that point in time. I'd check the memory region under gdb.

> i took the original code of the function and put a #[inline(never)] before ... code runs fine :D

Oh well...

**Comment by Arnold1** [comment]
bug fixed :D 

the following loop needs to be changed to for the release build (it works fin in debug), otherwise the code will crash with illegal instruction, reason: wrong code optimization: stmiacs r3!, {r1, r12} ... it stores two 32-bit values at a time...

use:
```
for i in 0..n {
              volatile_store::<u32>(&mut (*self.chaMem)[i], (*data)[i] as u32);
}
```
